src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/InputFlowControl.cs

```diff
@@ -19,94 +19,125 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.FlowControl;
             }
 
-            _flow.Advance(bytes);
-            return true;
-        }
+            computedFlow = new FlowControlState(currentFlow.Available - (uint)bytes, isAborted: false);
+        } while (currentFlow._state != Interlocked.CompareExchange(ref _flow._state, computedFlow._state, currentFlow._state));
+
+        return true;
     }
 
     public bool TryUpdateWindow(int bytes, out int updateSize)
     {
-        lock (_flowLock)
+        FlowControlState currentFlow, computedFlow;
+        do
         {
             updateSize = 0;
-
-            if (_flow.IsAborted)
+            currentFlow = _flow; // Copy
+            if (currentFlow.IsAborted)
             {
                 // All data received by stream has already been returned to the connection window.
                 return false;
             }
 
-            if (!_flow.TryUpdateWindow(bytes))
+            var maxUpdate = int.MaxValue - currentFlow.Available;
+            if (bytes > maxUpdate)
             {
                 // We only try to update the window back to its initial size after the app consumes data.
                 // It shouldn't be possible for the window size to ever exceed Http2PeerSettings.MaxWindowSize.
                 Debug.Assert(false, $"{nameof(TryUpdateWindow)} attempted to grow window past max size.");
             }
+            computedFlow = new FlowControlState(currentFlow.Available + (uint)bytes, isAborted: false);
+        } while (currentFlow._state != Interlocked.CompareExchange(ref _flow._state, computedFlow._state, currentFlow._state));
 
-            if (_windowUpdatesDisabled)
-            {
-                // Continue returning space to the connection window. The end of the stream has already
-                // been received, so don't send window updates for the stream window.
-                return true;
-            }
-
-            var potentialUpdateSize = _pendingUpdateSize + bytes;
+        if (_windowUpdatesDisabled)
+        {
+            // Continue returning space to the connection window. The end of the stream has already
+            // been received, so don't send window updates for the stream window.
+            return true;
+        }
 
+        int computedPendingUpdateSize, currentPendingSize;
+        do
+        {
+            updateSize = 0;
+            currentPendingSize = _pendingUpdateSize;
+            var potentialUpdateSize = currentPendingSize + bytes;
             if (potentialUpdateSize > _minWindowSizeIncrement)
             {
-                _pendingUpdateSize = 0;
+                computedPendingUpdateSize = 0;
                 updateSize = potentialUpdateSize;
             }
             else
             {
-                _pendingUpdateSize = potentialUpdateSize;
+                computedPendingUpdateSize = potentialUpdateSize;
             }
+        } while (currentPendingSize != Interlocked.CompareExchange(ref _pendingUpdateSize, computedPendingUpdateSize, currentPendingSize));
```

Comment:
In the locking version, the same thread that updated `_flow` updated `_pendingUpdateSize`; now it seems like different threads could update them.  It's not obvious to me whether that matters.

Reply:
As far as I can tell it does not, but I am not sure if I missed any use-case.